sdk/eventgrid/azure-messaging-eventgrid/src/main/java/com/azure/messaging/eventgrid/EventGridEvent.java

```diff
@@ -0,0 +1,291 @@
+                TypeReference.createInstance(SystemEventMappings.getSystemEventMappings().get(eventType)));
+        }
+        return event.getData();
+    }
+
+    /**
+     * Get the deserialized data property from the parsed event. The behavior is undefined if this method is called
+     * on an event that was not created through the parse method.
+     * @param clazz the class of the type to deserialize the data into.
+     * @param <T>   the type to deserialize the data into.
+     *
+     * @return the data deserialized into the given type using a default deserializer.
+     * @throws IllegalStateException If the event was not created through {@link EventGridEvent#parse(String)}.
+     */
+    public <T> T getData(Class<T> clazz) {
+        return getDataAsync(clazz, deserializer).block();
+    }
+
+    /**
+     * Get the deserialized data property from the parsed event.
+     * @param clazz the class of the type to deserialize the data into.
+     * @param <T>   the type to deserialize the data into.
+     *
+     * @return the data deserialized into the given type using a default deserializer, delivered asynchronously through
+     * a {@link Mono}.
+     * @throws IllegalStateException If the event was not created through {@link EventGridEvent#parse(String)}.
+     */
+    public <T> Mono<T> getDataAsync(Class<T> clazz) {
+        return getDataAsync(clazz, deserializer);
+    }
+
+    /**
+     * Get the deserialized data property from the parsed event.
+     * @param clazz            the class of the type to deserialize the data into.
+     * @param dataDeserializer the deserializer to use.
+     * @param <T>              the type to deserialize the data into.
+     *
+     * @return the data deserialized into the given type using the given deserializer.
+     * @throws IllegalStateException If the event was not created through {@link EventGridEvent#parse(String)}.
+     */
+    public <T> T getData(Class<T> clazz, JsonSerializer dataDeserializer) {
+        return getDataAsync(clazz, dataDeserializer).block();
+    }
+
+    /**
+     * Get the deserialized data property from the parsed event.
+     * @param clazz            the class of the type to deserialize the data into.
+     * @param dataDeserializer the deserializer to use.
+     * @param <T>              the type to deserialize the data into.
+     *
+     * @return the data deserialized into the given type using the given deserializer, delivered asynchronously through
+     * a {@link Mono}.
+     * @throws IllegalStateException If the event was not created through {@link EventGridEvent#parse(String)}.
+     */
+    public <T> Mono<T> getDataAsync(Class<T> clazz, JsonSerializer dataDeserializer) {
+        if (!parsed) {
+            // data was set instead of parsed, throw exception because we don't know how the data relates to clazz
+            throw logger.logExceptionAsError(new IllegalStateException(
```

Comment:
For async methods, exceptions should be passed through the publisher. Use `FluxUtil.monoError()`.